### PR TITLE
[SILGen] Handle arbitrary ObjC-to-Swift bridging for globals

### PR DESCRIPTION
**Pull request by jrose-apple**
Possibly useful for rdar://problem/34913634, possibly not. Now that it's working in my little test cases I need to go back and examine whether we actually need it, per discussion on #10832. May still be useful in some form for the SIL folks, though.

### CODE REVIEW

**Comment by slavapestov** [comment]
This looks good at first glance. It should obsolete my fix for dfdd3a795d6d6695c4920d0769730a325f480257 -- can you try removing it and see if the test still passes?

**Comment by slavapestov** [comment]
Actually my fix hits the code path for member VarDecls, but I think we should use the same bridging path component in both cases (the original bug was that NSString globals were bridged, but import-as-member globals were not).

**Review by rjmccall — COMMENTED** [review]
Looks good!

**Review by rjmccall — APPROVED** [review]

**Comment by jrose-apple** [comment]
Indeed, this obsoletes dfdd3a7. It *doesn't* seem to fix rdar://problem/34913634 after all, but now I feel comfortable committing it, since there's already a test case. I've folded in a revert of that change.

@swift-ci Please test

**Comment by jrose-apple** [comment]
Uh, hm. I wonder if the [Linux failure](https://ci.swift.org/job/swift-PR-Linux/1079/) is a fluke.

@swift-ci Please test Linux

File: lib/SILGen/SILGenLValue.cpp [hunk 1]
```diff
@@ -2219,6 +2280,12 @@ static LValue emitLValueForNonMemberVarDecl(SILGenFunction &SGF,
 
     if (address.getType().is<ReferenceStorageType>())
       lv.add<OwnershipComponent>(typeData);
+
+    AbstractionPattern pattern = SGF.SGM.Types.getAbstractionPattern(var);
+    if (pattern.isClangType()) {
```
Comment (jrose-apple) [inline lib/SILGen/SILGenLValue.cpp]: Hang on, should this also check that the types are different? What's the best way to do that? If they happen to be the same we might as well leave the physical component in place.

Comment (rjmccall) [reply lib/SILGen/SILGenLValue.cpp]: You want to check whether the type-of-rvalue matches the target lowered type.

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Review by slavapestov — APPROVED** [review]
Thanks for cleaning up my earlier fix too.

**Comment by DougGregor** [comment]
@swift-ci please smoke test macOS

**Comment by jrose-apple** [comment]
Had the wrong type for reverse bridging.

@swift-ci Please test

**Comment by jrose-apple** [comment]
I've tried several variations and can't get this right. Going to have to shelve it for now.

**Comment by jrose-apple** [comment]
Filed rdar://problem/35328468 to come back to this later.

**Pull request closed without merging**